packages/next/src/server/app-render/metadata-insertion/create-server-inserted-metadata.tsx

```diff
@@ -0,0 +1,42 @@
+import React from 'react'
+import { renderToReadableStream } from 'react-dom/server.edge'
+import {
+  ServerInsertedMetadataContext,
+  type MetadataResolver,
+} from '../../../shared/lib/server-inserted-metadata.shared-runtime'
+import { renderToString } from '../render-to-string'
+
+export function createServerInsertedMetadata() {
+  let metadataResolver: MetadataResolver | null = null
+  const setMetadataResolver = (resolver: MetadataResolver): void => {
+    metadataResolver = resolver
+  }
+
+  return {
+    ServerInsertedMetadataProvider: ({
+      children,
+    }: {
+      children: React.ReactNode
+    }) => {
+      return (
+        <ServerInsertedMetadataContext.Provider value={setMetadataResolver}>
+          {children}
+        </ServerInsertedMetadataContext.Provider>
+      )
+    },
+
+    async getServerInsertedMetadata(): Promise<string> {
```

Comment:
this function should only be able to return metadata at most once. Ideally it'd be called only once when streaming but because duplicating the head insertion logic is maybe not great we can at least ensure here within this function it can only ever produce the metadata value one time